### PR TITLE
Refactored workshop payment calculators, and moved into lib.

### PR DESCRIPTION
**Pull request by aoby**
1. Refactored the payment calculators again based on discussions earlier this week. Now the calculator emits a `WorkshopSummary`, which contains calculated workshop data (attendance #s, etc), a `WorkshopPayment` if it's qualified for payment, and a list of `TeacherSummary` objects that in turn contain `TeacherPayment` if that teacher is qualified for payment.
2. Moved payment & calculator classes from `app/models` into `lib`, since it does not have a DB backing.
3. Added calculator for Counselor/Admin workshops.

### CODE REVIEW

File: dashboard/lib/pd/payment/payment_factory.rb [hunk 1]
```diff
@@ -20,6 +20,11 @@ def self.get_calculator_class(workshop)
           Pd::Workshop::COURSE_CS_IN_A,
           Pd::Workshop::COURSE_CS_IN_S
         ].include?(workshop.course)
+
+        return PaymentCalculatorCounselorAdmin if [
```
Comment (mehalshah) [inline dashboard/lib/pd/payment/payment_factory.rb:24]: Why are these the same?

Comment (mehalshah) [reply dashboard/lib/pd/payment/payment_factory.rb:24]: Nm, this is fine

File: dashboard/lib/pd/payment/payment_calculator_base.rb [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+          end
+        end
+      end
+    end
+
+    # Constructs workshop payment from a workshop summary
+    # @param workshop_summary [WorkshopSummary]
+    # @return [WorkshopPayment, nil] Payment details, or nil if the workshop is not qualified for payment.
+    def construct_workshop_payment(workshop_summary)
+      return nil unless qualified? workshop_summary.total_teacher_attendance_days
```
Comment (mehalshah) [inline dashboard/lib/pd/payment/payment_calculator_base.rb]: I think standard is to put parens around this

File: dashboard/lib/pd/payment/payment_calculator_base.rb [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+      # Generate a teacher summary for all teachers in raw attendance.
+      raw_teacher_attendance.keys.map do |teacher_id|
+        enrollment = enrollments_by_teacher_id[teacher_id]
+        teacher = enrolled_teachers_by_id[teacher_id] || User.find(teacher_id)
+
+        TeacherSummary.new.tap do |teacher_summary|
+          raw_attendance = raw_teacher_attendance[teacher_id]
+
+          teacher_summary.workshop_summary = workshop_summary
```
Comment (mehalshah) [inline dashboard/lib/pd/payment/payment_calculator_base.rb]: I can't help but feel like there's a better pattern to write these statements - I worry about this class getting bloated with boilerplate logic.

You could always just have a create method on teacher summaries that takes in parameters

File: dashboard/test/lib/pd/payment/payment_calculator_base_test.rb [hunk 4]
```diff
@@ -0,0 +1,211 @@
+require 'test_helper'
```
Comment (mehalshah) [inline dashboard/test/lib/pd/payment/payment_calculator_base_test.rb:1]: When you look in code coverage, is all of this covered from what you can see?

File: dashboard/lib/pd/payment/payment_calculator_base.rb (hunk 2 above)
Comment (aoby) [reply dashboard/lib/pd/payment/payment_calculator_base.rb]: fixed

File: dashboard/lib/pd/payment/payment_calculator_base.rb (hunk 3 above)
Comment (aoby) [reply dashboard/lib/pd/payment/payment_calculator_base.rb]: Good call. I wrote constructors for `WorkshopSummary`, `WorkshopPayment`, `TeacherSummary`, and `TeacherPayment` each with required named parameters and made those fields read-only.

File: dashboard/test/lib/pd/payment/payment_calculator_base_test.rb (hunk 4 above)
Comment (aoby) [reply dashboard/test/lib/pd/payment/payment_calculator_base_test.rb:1]: Now it's 100% on the calculator classes.

**Pull request merged**